src/datamodel/quests/quest_new.rs

```diff
@@ -1,64 +1,45 @@
-use std::collections::HashMap;
-use std::rc::Rc;
+use std::hash::Hash;
 
-#[derive(Debug, PartialEq)]
+use crate::core::election::Election;
+
+#[derive(Hash, PartialEq, Eq, Clone, Debug, Copy)]
 pub enum Vote {
     Success,
     Failed,
 }
 
-pub struct QuestObjective {
-    pub num_players: u32,
-    pub required_successes: u32,
-}
-
-pub struct QuestNew {
-    members: HashMap<&'static str, bool>,
-    objective: Rc<QuestObjective>,
-    success_votes: u32,
+pub struct QuestNew<'a> {
+    election: Election<Vote>,
+    winner_rule: &'a (Vote, usize),
 }
 
-impl QuestNew {
-    pub fn new(members: Vec<&'static str>, objective: Rc<QuestObjective>) -> QuestNew {
-        let members: HashMap<&'static str, bool> =
-            members.into_iter().map(|m| (m, false)).collect();
-        QuestNew {
-            objective,
-            members,
-            success_votes: 0,
-        }
+impl<'a> QuestNew<'a> {
+    pub fn new(quest_member: &[String], winner_rule: &'a (Vote, usize)) -> QuestNew<'a> {
+        assert!(quest_member.len() >= winner_rule.1, "Winner's rule cannot be bigger than quest member number");
+        let election = Election::<Vote>::new(quest_member);
+        QuestNew { election, winner_rule, }
     }
 
-    pub fn add_vote(&mut self, player: &str, vote: Vote) {
-        if self.members[player] == true {
-            return;
-        }
-        if vote == Vote::Success {
-            self.success_votes += 1;
-        }
-        *self.members.get_mut(player).unwrap() = true;
+    pub fn vote(&mut self, quest_member: &String, vote: Vote) {
+        self.election.vote(quest_member, vote);
     }
 
-    pub fn finish_quest(self) -> Result<FinishedQuestNew, QuestNew> {
-        let unvoted: Vec<bool> = self.members.iter().map(|m| *m.1).filter(|v| !*v).collect();
-        if unvoted.len() > 0 {
-            Err(self)
-        } else {
-            if self.success_votes == self.objective.required_successes {
-                Ok(FinishedQuestNew { success: true })
-            } else {
-                Ok(FinishedQuestNew { success: false })
-            }
+    pub fn finish_quest(self) -> QuestResult {
+        let scrutiny = self.election.count_votes().ok().unwrap();
+        match (self.winner_rule.0 , scrutiny.result().get(&Some(self.winner_rule.0))) { 
+            (_ , Some(&votes_num)) if votes_num >= self.winner_rule.1 => QuestResult{ quest_result: self.winner_rule.0, },
+            (Vote::Success , _) => QuestResult{ quest_result: Vote::Failed, },
+            (Vote::Failed , _) => QuestResult{ quest_result: Vote::Success, },
```

Comment:
Maybe it is just that my Rust level is still low, but I have some trouble understanding what this function is actually doing. I feel we might be getting into an unnecessarily complex level, too general for the use case we have. Couldn't this be simplified so it is more readable (maybe less optimal or compact)?

In general one should first aim for correctness, then readability and finally optimisation/compactness, or Paraphrasing Don Knuth, "Premature optimisation is the root of all evil", so I just want to make sure this makes sense.

Reply:
haha This match is a master piece. In fact, I'm using lots of concepts here but most are in this [section of the book](https://doc.rust-lang.org/book/ch18-01-all-the-places-for-patterns.html).

Reply:
I guess there are several implicit aspects within this expression:
1. The easiest one: `match` is returning a `QuestResult` and that's why there isn't a `;` after the closing bracket. The `return` operator is implicit there (very _rustacianistic_ ;-) ).
2. Why using a tuple to evaluate the cases? Because we have 2 possible scenarios. The fist one, we get a winner according to the `winner_rule` and we return it. Second scenario is where the winner rule isn't reached, the candidate hasn't gotten the votes to be a winner. Then, we'll return the opposite candidate. If it's `Success` we'll return `Failed` and the same with `Failed`. I've used the [Ignoring operator ](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-parts-of-a-value-with-a-nested-_) to switch between cases. 
3. The first arm of the branch might be the most complex because it uses [Match Guards conditonals](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#extra-conditionals-with-match-guards) and also [Irrefutable pattern](https://doc.rust-lang.org/book/ch18-02-refutability.html#refutability-whether-a-pattern-might-fail-to-match).

Reply:
Cargo fmt made the `match` less readable I think, nevertheless I still think this is the most rustatian way.

Reply:
See my comment in the next review.